src/ray/raylet/worker_pool.cc

```diff
@@ -863,14 +869,39 @@ std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
       if (task_spec.IsActorCreationTask()) {
         dynamic_options = task_spec.DynamicWorkerOptions();
       }
-      proc = StartWorkerProcess(task_spec.GetLanguage(), rpc::WorkerType::WORKER,
-                                task_spec.JobId(), dynamic_options,
-                                task_spec.SerializedRuntimeEnv(),
-                                task_spec.OverrideEnvironmentVariables());
-      if (proc.IsValid()) {
-        state.pending_dedicated_workers_to_tasks[proc] = task_spec.TaskId();
-        state.tasks_to_pending_dedicated_workers[task_spec.TaskId()] = proc;
+      auto start_worker_process_fn =
+          [this](const TaskSpecification &task_spec, State &state,
+                 std::vector<std::string> dynamic_options) -> Process {
+        Process proc = StartWorkerProcess(
+            task_spec.GetLanguage(), rpc::WorkerType::WORKER, task_spec.JobId(),
+            dynamic_options, task_spec.SerializedRuntimeEnv(),
+            task_spec.OverrideEnvironmentVariables());
+        if (proc.IsValid()) {
+          state.pending_dedicated_workers_to_tasks[proc] = task_spec.TaskId();
+          state.tasks_to_pending_dedicated_workers[task_spec.TaskId()] = proc;
+        }
+        return proc;
+      };
+
+      // create runtime env.
+      if (!(task_spec.SerializedRuntimeEnv() == "{}" ||
+            task_spec.SerializedRuntimeEnv() == "")) {
+        create_runtime_env = true;
+        state.tasks_to_pending_runtime_envs[task_spec.TaskId()] =
+            task_spec.SerializedRuntimeEnv();
+        agent_manager_->CreateRuntimeEnvOrReuse(
+            task_spec.SerializedRuntimeEnv(),
+            [start_worker_process_fn, &state, task_spec, dynamic_options](
+                Status status, const rpc::CreateRuntimeEnvReply &reply) {
+              if (!status.ok()) {
+                RAY_LOG(ERROR)
+                    << "Create runtime env rpc failed. Wait for next time to retry.";
+                return;
```

Comment:
This will be a pretty bad user experience if there is an error in the environment that causes it to continually fail. Ideally, we should optionally retry some number of times and then fail the tasks that depend on the environment with an `EnvironmentFailureError` or something

Reply:
@edoakes Yes, I have add some retry code.